s3fs/core.py

```diff
@@ -2331,41 +2333,41 @@ def _upload_chunk(self, final=False):
             and self.tell() < self.blocksize
         ):
             # only happens when closing small file, use on-shot PUT
-            data1 = False
+            pass
         else:
             self.buffer.seek(0)
-            (data0, data1) = (None, self.buffer.read(self.blocksize))
-
-        while data1:
-            (data0, data1) = (data1, self.buffer.read(self.blocksize))
-            data1_size = len(data1)
 
-            if 0 < data1_size < self.blocksize:
-                remainder = data0 + data1
-                remainder_size = self.blocksize + data1_size
+            def upload_part(part_data: bytes):
+                if len(part_data) == 0:
+                    return
+                part = len(self.parts) + 1
+                logger.debug("Upload chunk %s, %s" % (self, part))
 
-                if remainder_size <= self.part_max:
-                    (data0, data1) = (remainder, None)
-                else:
-                    partition = remainder_size // 2
-                    (data0, data1) = (remainder[:partition], remainder[partition:])
+                out = self._call_s3(
+                    "upload_part",
+                    Bucket=bucket,
+                    PartNumber=part,
+                    UploadId=self.mpu["UploadId"],
+                    Body=part_data,
+                    Key=key,
+                )
 
-            part = len(self.parts) + 1
-            logger.debug("Upload chunk %s, %s" % (self, part))
+                part_header = {"PartNumber": part, "ETag": out["ETag"]}
+                if "ChecksumSHA256" in out:
+                    part_header["ChecksumSHA256"] = out["ChecksumSHA256"]
+                self.parts.append(part_header)
 
-            out = self._call_s3(
-                "upload_part",
-                Bucket=bucket,
-                PartNumber=part,
-                UploadId=self.mpu["UploadId"],
-                Body=data0,
-                Key=key,
-            )
+            def n_bytes_left() -> int:
+                return len(self.buffer.getbuffer()) - self.buffer.tell()
 
-            part_header = {"PartNumber": part, "ETag": out["ETag"]}
-            if "ChecksumSHA256" in out:
-                part_header["ChecksumSHA256"] = out["ChecksumSHA256"]
-            self.parts.append(part_header)
+            min_chunk = 1 if final else self.blocksize
+            if self.fs.fixed_upload_size:
+                # all chunks have fixed size, exception: last one can be smaller
+                while n_bytes_left() >= min_chunk:
```

Comment:
```suggestion
                while n_bytes_left() >= self.blocksize:
```
?

Reply:
won't cover last chunk if final=True

Reply:
But it *will* include last part of the buffer, with size < block_size on each call to flush(), so if blocksize is 10000, 
```
f.write(11000) # uploads 10000 and 1000
f.write(11000) # fails, block size mismatch because of preceding 1000-byte block
```

Reply:
unless I'm missing something, in you example:

```
f.write(11000)
# final = false -> min_chunk=10_000, uploads 10_000 and lefts 1_000 in buffer
f.write(11000)
# final = True (for example) -> min_chunk=1, uploads 10_000 and then final chunk 2_000

```

Reply:
Lets make a test that does exactly this (with fixed_upload_size=True) and make sure it passes, and 1000 is in the buffer between the two calls.

Reply:
@martindurant  wrote, test fails. 

Now I'm lost about your previous messages - you recommended not reading last chunk, and I was under impression that it is carried over, but apparently it isn't.

Does it mean I should go back to implementation that adds a buffer to s3file to keep data that wasn't sent?

Reply:
The file object has a buffer; the return value of the upload method tells the caller whether to empty the buffer or not. In your case, you probably want to _replace_ the buffer you just read from with whatever was not yet sent because it's not big enough.

Reply:
Let me know if you need more information about this.